FhirEtl with DataflowRunner does not produce Parquet files

I run a command like this: 

`java -cp ./pipelines/batch/target/batch-bundled.jar com.google.fhir.analytics.FhirEtl --outputParquetPath=gs://fhir-analytics-test/parquet_from-json_79_30workers --runner=DataflowRunner --sourceJsonFilePattern=gs://synthea-hiv/79_patients/* --resourceList="Patient,Encounter,Observation" --project=fhir-analytics-test --region=us-central1 --gcpTempLocation=gs://fhir-analytics-test/dataflow_temp --numWorkers=30 
`
It takes about 5 minutes and ends with success. But the output path only contains timestamp_start/end.txt and no Parquet files.

This appears on the console:

21:43:45.091 [main] INFO  com.google.fhir.analytics.FhirEtl com.google.fhir.analytics.FhirEtl.main:443 - Flags: Current Settings:
  appName: FhirEtl
  fhirServerUrl: 
  fhirSinkPath: 
  gcpTempLocation: gs://fhir-analytics-test/dataflow_temp
  numWorkers: 30
  optionsId: 0
  outputParquetPath: gs://fhir-analytics-test/parquet_from-json_79_30workers_20240416
  parquetInputDwhRoot: 
  project: fhir-analytics-test
  region: us-central1
  resourceList: Patient,Encounter,Observation
  runner: class org.apache.beam.runners.dataflow.DataflowRunner
  sourceJsonFilePattern: gs://synthea-hiv/79_patients/*
  stableUniqueNames: WARNING

21:43:45.105 [main] INFO  ca.uhn.fhir.util.VersionUtil ca.uhn.fhir.util.VersionUtil.initialize:84 - HAPI FHIR version 6.10.4 - Rev 5d6bb47853
21:43:45.108 [main] INFO  ca.uhn.fhir.context.FhirContext ca.uhn.fhir.context.FhirContext.<init>:226 - Creating new FHIR context for FHIR version [R4]
21:43:45.132 [main] INFO  c.u.f.c.s.DefaultProfileValidationSupport c.u.f.c.s.DefaultProfileValidationSupportBundleStrategy.loadStructureDefinitions:368 - Loading structure definitions from classpath: /org/hl7/fhir/r4/model/profile/profiles-resources.xml
21:43:45.144 [main] INFO  ca.uhn.fhir.util.XmlUtil ca.uhn.fhir.util.jar.DependencyLogImpl.logStaxImplementation:76 - FHIR XML procesing will use StAX implementation 'null' version 'null'
21:43:45.904 [main] INFO  c.u.f.c.s.DefaultProfileValidationSupport c.u.f.c.s.DefaultProfileValidationSupportBundleStrategy.loadStructureDefinitions:368 - Loading structure definitions from classpath: /org/hl7/fhir/r4/model/profile/profiles-types.xml
21:43:45.946 [main] INFO  c.u.f.c.s.DefaultProfileValidationSupport c.u.f.c.s.DefaultProfileValidationSupportBundleStrategy.loadStructureDefinitions:368 - Loading structure definitions from classpath: /org/hl7/fhir/r4/model/profile/profiles-others.xml
21:43:46.036 [main] INFO  c.u.f.c.s.DefaultProfileValidationSupport c.u.f.c.s.DefaultProfileValidationSupportBundleStrategy.loadStructureDefinitions:368 - Loading structure definitions from classpath: /org/hl7/fhir/r4/model/extension/extension-definitions.xml
Apr 16, 2024 9:43:46 PM com.google.auth.oauth2.DefaultCredentialsProvider warnAboutProblematicCredentials
WARNING: You are authenticating using user credentials. For production, we recommend using service account credentials.

To learn more about service account credentials, see http://cloud.google.com/docs/authentication/external/set-up-adc-on-cloud
21:49:18.943 [pool-6-thread-1] INFO  com.google.fhir.analytics.EtlUtils com.google.fhir.analytics.EtlUtils.logMetrics:56 - Pipeline counter PipelineMetrics:numFetchedResources_ReadJsonFiles : 46497
21:49:18.944 [pool-6-thread-1] INFO  com.google.fhir.analytics.EtlUtils com.google.fhir.analytics.EtlUtils.logMetrics:56 - Pipeline counter PipelineMetrics:totalGenerateTimeMillis_ReadJsonFiles : 25394
21:49:19.250 [main] INFO  com.google.fhir.analytics.FhirEtl com.google.fhir.analytics.FhirEtl.main:454 - DONE!

DirectRunner fails and outputs some files into GCP storage.

export RUNNER=Direct; java -cp ./pipelines/batch/target/batch-bundled.jar com.google.fhir.analytics.FhirEtl --outputParquetPath=gs://fhir-analytics-test/parquet_from-json_79_$( printf '%(%Y-%m-%d_%H:%M:%S)T' )_$RUNNER --runner=${RUNNER}Runner --sourceJsonFilePattern=gs://synthea-hiv/79_patients/* --resourceList="Patient,Encounter,Observation" --project=fhir-analytics-test --region=us-central1 --gcpTempLocation=gs://fhir-analytics-test/dataflow_temp --numWorkers=1

14:44:32.228 [direct-runner-worker] INFO  c.g.fhir.analytics.ReadJsonFilesFn com.google.fhir.analytics.ReadJsonFilesFn.processElement:50 - Reading file with metadata Metadata{resourceId=gs://synthea-hiv/79_patients/Kelley882_Marks830_898c1dc8-2c3d-7afa-5e61-eeadba8f7864.json, sizeBytes=871851, isReadSeekEfficient=true, checksum=null, lastModifiedMillis=0}
14:44:35.111 [direct-runner-worker] ERROR o.a.b.r.d.TransformEvaluatorRegistry o.a.beam.runners.direct.TransformEvaluatorRegistry.cleanup:179 - Exceptions [java.lang.Exception: Exceptions thrown while tearing down DoFns] thrown while cleaning up evaluators
14:44:35.112 [main] ERROR com.google.fhir.analytics.EtlUtils com.google.fhir.analytics.EtlUtils.runMultiplePipelines:141 - Error while executing the pipelines
java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Error during executor shutdown:
- Exceptions thrown while tearing down DoFns
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at com.google.fhir.analytics.EtlUtils.runMultiplePipelines(EtlUtils.java:129)
	at com.google.fhir.analytics.EtlUtils.runMultiplePipelinesWithTimestamp(EtlUtils.java:78)
	at com.google.fhir.analytics.FhirEtl.main(FhirEtl.java:451)
Caused by: java.lang.IllegalStateException: Error during executor shutdown:
- Exceptions thrown while tearing down DoFns
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor.shutdownIfNecessary(ExecutorServiceParallelExecutor.java:349)
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor.access$200(ExecutorServiceParallelExecutor.java:63)
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor$2.run(ExecutorServiceParallelExecutor.java:193)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Exception in thread "main" java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Error during executor shutdown:
- Exceptions thrown while tearing down DoFns
	at com.google.fhir.analytics.EtlUtils.runMultiplePipelines(EtlUtils.java:142)
	at com.google.fhir.analytics.EtlUtils.runMultiplePipelinesWithTimestamp(EtlUtils.java:78)
	at com.google.fhir.analytics.FhirEtl.main(FhirEtl.java:451)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Error during executor shutdown:
- Exceptions thrown while tearing down DoFns
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at com.google.fhir.analytics.EtlUtils.runMultiplePipelines(EtlUtils.java:129)
	... 2 more
Caused by: java.lang.IllegalStateException: Error during executor shutdown:
- Exceptions thrown while tearing down DoFns
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor.shutdownIfNecessary(ExecutorServiceParallelExecutor.java:349)
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor.access$200(ExecutorServiceParallelExecutor.java:63)
	at org.apache.beam.runners.direct.ExecutorServiceParallelExecutor$2.run(ExecutorServiceParallelExecutor.java:193)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)

--- a/pipelines/batch/src/main/java/com/google/fhir/analytics/ConvertResourceFn.java
+++ b/pipelines/batch/src/main/java/com/google/fhir/analytics/ConvertResourceFn.java
@@ -192,6 +192,11 @@ public class ConvertResourceFn extends FetchSearchPageFn<HapiRowDescriptor> {
     }
   }
 
+  @Override
+  public void finishBundle(FinishBundleContext context) {
+    super.finishBundle(context);
+  }
+
   @ProcessElement
   public void processElement(ProcessContext processContext)
       throws IOException, ParseException, SQLException, ViewApplicationException, ProfileException {

--- a/pipelines/batch/src/main/java/com/google/fhir/analytics/FetchPatientHistory.java
+++ b/pipelines/batch/src/main/java/com/google/fhir/analytics/FetchPatientHistory.java
@@ -53,6 +53,10 @@ public class FetchPatientHistory extends PTransform<PCollection<KV<String, Integ
 
     fetchSearchPageFn =
         new FetchSearchPageFn<KV<String, Integer>>(options, stageId) {
+          @Override
+          public void finishBundle(FinishBundleContext context) {
+            super.finishBundle(context);
+          }
 
           @ProcessElement
           public void ProcessElement(@Element KV<String, Integer> patientIdCount)

--- a/pipelines/batch/src/main/java/com/google/fhir/analytics/FetchPatients.java
+++ b/pipelines/batch/src/main/java/com/google/fhir/analytics/FetchPatients.java
@@ -52,6 +52,11 @@ public class FetchPatients extends PTransform<PCollection<KV<String, Integer>>, 
     fetchSearchPageFn =
         new FetchSearchPageFn<KV<String, Integer>>(options, "PatientById") {
 
+          @Override
+          public void finishBundle(FinishBundleContext context) {
+            super.finishBundle(context);
+          }
+
           @ProcessElement
           public void processElement(@Element KV<String, Integer> patientIdCount)
               throws IOException, SQLException, ViewApplicationException, ProfileException {

--- a/pipelines/batch/src/main/java/com/google/fhir/analytics/FetchResources.java
+++ b/pipelines/batch/src/main/java/com/google/fhir/analytics/FetchResources.java
@@ -63,6 +63,8 @@ public class FetchResources
   @VisibleForTesting
   static String getSubjectPatientIdOrNull(Resource resource) {
     String patientId = null;
+    // TODO: Instead of hard-coding `subject` here, use Patient Compartment:
+    //  https://hl7.org/fhir/compartmentdefinition-patient.html
     Property subject = resource.getNamedProperty("subject");
     if (subject != null) {
       List<Base> values = subject.getValues();
@@ -119,12 +121,13 @@ public class FetchResources
       patientCount.put(patientId, current + 1);
     }
 
-    @FinishBundle
+    @Override
     public void finishBundle(FinishBundleContext context) {
       for (Map.Entry<String, Integer> entry : patientCount.entrySet()) {
         context.output(
             KV.of(entry.getKey(), entry.getValue()), Instant.now(), GlobalWindow.INSTANCE);
       }
+      super.finishBundle(context);
     }
 
     @ProcessElement

--- a/pipelines/batch/src/main/java/com/google/fhir/analytics/FetchSearchPageFn.java
+++ b/pipelines/batch/src/main/java/com/google/fhir/analytics/FetchSearchPageFn.java
@@ -110,9 +110,9 @@ abstract class FetchSearchPageFn<T> extends DoFn<T, KV<String, Integer>> {
 
   protected IParser parser;
 
-  private FhirVersionEnum fhirVersionEnum;
+  private final FhirVersionEnum fhirVersionEnum;
 
-  private String structureDefinitionsPath;
+  private final String structureDefinitionsPath;
 
   protected AvroConversionUtil avroConversionUtil;
 

--- a/pipelines/batch/src/main/java/com/google/fhir/analytics/FetchSearchPageFn.java
+++ b/pipelines/batch/src/main/java/com/google/fhir/analytics/FetchSearchPageFn.java
@@ -45,7 +45,10 @@ import org.slf4j.LoggerFactory;
  * to Avro and JSON records. The non-abstract sub-classes should implement `ProcessElement` using
  * `processBundle` auxiliary method. Note the code reuse pattern that we really need here is
  * composition (not inheritance) but because of Beam complexities (e.g., certain work need to be
- * done during `setup()` where PipelienOptions not available) we use inheritance.
+ * done during `setup()` where PipelienOptions not available) we use inheritance. A better approach
+ * is to create the utility instances (e.g., `fetchUtil`) once at the beginning of ParDo or
+ * StartBundle method using a synchronized method. Those functions have acccess to PipelienOptions.
+ * That way we can get rid of many instance variables that mirror PipelienOptions fields.
  *
  * @param <T> The type of the elements of the input PCollection.
  */
@@ -215,8 +218,30 @@ abstract class FetchSearchPageFn<T> extends DoFn<T, KV<String, Integer>> {
     }
   }
 
+  /**
+   * This should be overridden by all subclasses because the `context` type is not fully specified
+   * at this parent class (because of the T type argument). All subclass implementations should call
+   * `super.finishBundle` though. TODO: implement a way to enforce this at compile time; this is
+   * currently caught at run time.
+   */
+  @FinishBundle
+  public void finishBundle(FinishBundleContext context) {
+    try {
+      if (parquetUtil != null) {
+        parquetUtil.flushAll();
+      }
+    } catch (IOException | ProfileException e) {
+      // There is not much that we can do at finishBundle so just throw a RuntimeException
+      log.error("At finishBundle caught exception ", e);
+      throw new IllegalStateException(e);
+    }
+  }
+
   @Teardown
   public void teardown() throws IOException {
+    // Note this is _not_ guaranteed to be called; for example when the worker process is being
+    // stopped, the runner may choose not to call teardown. Only keep closing/cleanups here. See:
+    // https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/transforms/DoFn.Teardown.html
     if (parquetUtil != null) {
       parquetUtil.closeAllWriters();
     }

--- a/pipelines/batch/src/main/java/com/google/fhir/analytics/FhirEtl.java
+++ b/pipelines/batch/src/main/java/com/google/fhir/analytics/FhirEtl.java
@@ -26,7 +26,6 @@ import com.google.fhir.analytics.metrics.PipelineMetrics;
 import com.google.fhir.analytics.metrics.PipelineMetricsProvider;
 import com.google.fhir.analytics.model.DatabaseConfiguration;
 import com.google.fhir.analytics.view.ViewDefinitionException;
-import java.beans.PropertyVetoException;
 import java.io.IOException;
 import java.sql.SQLException;
 import java.util.ArrayList;
@@ -183,7 +182,7 @@ public class FhirEtl {
 
   private static List<Pipeline> buildOpenmrsJdbcPipeline(
       FhirEtlOptions options, AvroConversionUtil avroConversionUtil)
-      throws PropertyVetoException, IOException, SQLException, ProfileException {
+      throws IOException, SQLException, ProfileException {
     // TODO add incremental support.
     Preconditions.checkArgument(Strings.isNullOrEmpty(options.getSince()));
     FhirSearchUtil fhirSearchUtil =
@@ -310,7 +309,7 @@ public class FhirEtl {
 
   // TODO: Implement active period feature for JDBC mode with a HAPI source server (issue #278).
   private static List<Pipeline> buildHapiJdbcPipeline(FhirEtlOptions options)
-      throws PropertyVetoException, SQLException, IOException {
+      throws SQLException, IOException {
     Preconditions.checkArgument(!Strings.isNullOrEmpty(options.getFhirDatabaseConfigPath()));
     DatabaseConfiguration dbConfig =
         DatabaseConfiguration.createConfigFromFile(options.getFhirDatabaseConfigPath());
@@ -429,8 +428,7 @@ public class FhirEtl {
    */
   static List<Pipeline> setupAndBuildPipelines(
       FhirEtlOptions options, AvroConversionUtil avroConversionUtil)
-      throws PropertyVetoException, IOException, SQLException, ViewDefinitionException,
-          ProfileException {
+      throws IOException, SQLException, ViewDefinitionException, ProfileException {
     if (!options.getSinkDbConfigPath().isEmpty()) {
       JdbcResourceWriter.createTables(options);
     }
@@ -450,8 +448,7 @@ public class FhirEtl {
   }
 
   public static void main(String[] args)
-      throws PropertyVetoException, IOException, SQLException, ViewDefinitionException,
-          ProfileException {
+      throws IOException, SQLException, ViewDefinitionException, ProfileException {
 
     AvroConversionUtil.initializeAvroConverters();
 

--- a/pipelines/batch/src/main/java/com/google/fhir/analytics/JdbcResourceWriter.java
+++ b/pipelines/batch/src/main/java/com/google/fhir/analytics/JdbcResourceWriter.java
@@ -30,7 +30,6 @@ import com.google.fhir.analytics.view.ViewDefinition;
 import com.google.fhir.analytics.view.ViewDefinitionException;
 import com.google.fhir.analytics.view.ViewManager;
 import com.google.fhir.analytics.view.ViewSchema;
-import java.beans.PropertyVetoException;
 import java.io.IOException;
 import java.sql.Connection;
 import java.sql.JDBCType;
@@ -107,7 +106,7 @@ public class JdbcResourceWriter {
   }
 
   static void createTables(FhirEtlOptions options)
-      throws PropertyVetoException, IOException, SQLException, ViewDefinitionException {
+      throws IOException, SQLException, ViewDefinitionException {
     // This should not be triggered in pipeline workers because concurrent CREATEs lead to failures:
     // https://stackoverflow.com/questions/54351783/duplicate-key-value-violates-unique-constraint
     Preconditions.checkArgument(!Strings.nullToEmpty(options.getSinkDbConfigPath()).isEmpty());

--- a/pipelines/batch/src/main/java/com/google/fhir/analytics/ProcessGenericRecords.java
+++ b/pipelines/batch/src/main/java/com/google/fhir/analytics/ProcessGenericRecords.java
@@ -64,17 +64,17 @@ public class ProcessGenericRecords extends FetchSearchPageFn<GenericRecord> {
   }
 
   @Override
-  public void teardown() throws IOException {
+  public void finishBundle(FinishBundleContext context) {
     if (!cachedResources.isEmpty()) {
       try {
         processBundle(flushCachToBundle());
-      } catch (SQLException | ViewApplicationException | ProfileException e) {
-        // This is not perfect but the parent teardown only has IOException.
-        log.error("Caught exception in teardown: ", e);
-        throw new IOException(e);
+      } catch (SQLException | ViewApplicationException | ProfileException | IOException e) {
+        // We cannot do much at finishBundle!
+        log.error("At finishBundle caught exception ", e);
+        throw new IllegalStateException(e);
       }
     }
-    super.teardown();
+    super.finishBundle(context);
   }
 
   private Bundle flushCachToBundle() {

--- a/pipelines/batch/src/main/java/com/google/fhir/analytics/ReadJsonFilesFn.java
+++ b/pipelines/batch/src/main/java/com/google/fhir/analytics/ReadJsonFilesFn.java
@@ -57,6 +57,11 @@ public class ReadJsonFilesFn extends FetchSearchPageFn<FileIO.ReadableFile> {
     }
   }
 
+  @Override
+  public void finishBundle(FinishBundleContext context) {
+    super.finishBundle(context);
+  }
+
   @ProcessElement
   public void processElement(@Element FileIO.ReadableFile file)
       throws IOException, SQLException, ViewApplicationException, ProfileException {

--- a/pipelines/common/src/main/java/com/google/fhir/analytics/ParquetUtil.java
+++ b/pipelines/common/src/main/java/com/google/fhir/analytics/ParquetUtil.java
@@ -103,6 +103,9 @@ public class ParquetUtil {
             @Override
             public void run() {
               try {
+                // If a flush has happened recently, e.g., through a `FinishBundle` call, we don't
+                // need to do that again; the timer here is just to make sure the data is flushed
+                // into Parquet files _at least_ once in every `secondsToFlush`.
                 if (!flushedInCurrentPeriod) {
                   log.info("Flush timed out for thread " + Thread.currentThread().getId());
                   flushAll();

--- a/pipelines/common/src/main/java/com/google/fhir/analytics/ParquetUtil.java
+++ b/pipelines/common/src/main/java/com/google/fhir/analytics/ParquetUtil.java
@@ -52,6 +52,11 @@ public class ParquetUtil {
   private final Timer timer;
   private final Random random;
   private final String namePrefix;
+  private boolean flushedInCurrentPeriod;
+
+  private synchronized void setFlushedInCurrentPeriod(boolean value) {
+    flushedInCurrentPeriod = value;
+  }
 
   public String getParquetPath() {
     return dwhFiles.getRoot();
@@ -90,6 +95,7 @@ public class ParquetUtil {
     this.writerMap = new HashMap<>();
     this.rowGroupSize = rowGroupSize;
     this.namePrefix = namePrefix;
+    setFlushedInCurrentPeriod(false);
     if (secondsToFlush > 0) {
       TimerTask task =
           new TimerTask() {
@@ -97,9 +103,11 @@ public class ParquetUtil {
             @Override
             public void run() {
               try {
-                log.info(
-                    "Flushing all Parquet writers for thread " + Thread.currentThread().getId());
-                flushAll();
+                if (!flushedInCurrentPeriod) {
+                  log.info("Flush timed out for thread " + Thread.currentThread().getId());
+                  flushAll();
+                }
+                setFlushedInCurrentPeriod(false);
               } catch (IOException | ProfileException e) {
                 log.error("Could not flush Parquet files: " + e);
               }
@@ -174,10 +182,12 @@ public class ParquetUtil {
     }
   }
 
-  private synchronized void flushAll() throws IOException, ProfileException {
+  synchronized void flushAll() throws IOException, ProfileException {
+    log.info("Flushing all Parquet writers for thread " + Thread.currentThread().getId());
     for (String resourceType : writerMap.keySet()) {
       flush(resourceType);
     }
+    setFlushedInCurrentPeriod(true);
   }
 
   public synchronized void closeAllWriters() throws IOException {

